packages/protocol/engine/congestioncontrol/icca/scheduler/scheduler.go

```diff
@@ -0,0 +1,610 @@
+	s.bufferMutex.RLock()
+	defer s.bufferMutex.RUnlock()
+
+	issuerQueue := s.buffer.IssuerQueue(issuerID)
+	if issuerQueue == nil {
+		return 0
+	}
+	return issuerQueue.Size()
+}
+
+// IssuerQueueSizes returns the size for each issuer queue.
+func (s *Scheduler) IssuerQueueSizes() map[identity.ID]int {
+	s.EvictionManager.RLock()
+	defer s.EvictionManager.RUnlock()
+	s.bufferMutex.RLock()
+	defer s.bufferMutex.RUnlock()
+
+	queueSizes := make(map[identity.ID]int)
+	for _, issuerID := range s.buffer.IssuerIDs() {
+		size := s.buffer.IssuerQueue(issuerID).Size()
+		queueSizes[issuerID] = size
+	}
+	return queueSizes
+}
+
+// MaxBufferSize returns the max size of the buffer.
+func (s *Scheduler) MaxBufferSize() int {
+	s.EvictionManager.RLock()
+	defer s.EvictionManager.RUnlock()
+	s.bufferMutex.RLock()
+	defer s.bufferMutex.RUnlock()
+
+	return s.buffer.MaxSize()
+}
+
+// BufferSize returns the size of the buffer.
+func (s *Scheduler) BufferSize() int {
+	s.EvictionManager.RLock()
+	defer s.EvictionManager.RUnlock()
+	s.bufferMutex.RLock()
+	defer s.bufferMutex.RUnlock()
+
+	return s.buffer.Size()
+}
+
+// ReadyBlocksCount returns the size buffer.
+func (s *Scheduler) ReadyBlocksCount() int {
+	s.EvictionManager.RLock()
+	defer s.EvictionManager.RUnlock()
+	s.bufferMutex.RLock()
+	defer s.bufferMutex.RUnlock()
+
+	return s.buffer.ReadyBlocksCount()
+}
+
+// TotalBlocksCount returns the size buffer.
+func (s *Scheduler) TotalBlocksCount() int {
+	s.EvictionManager.RLock()
+	defer s.EvictionManager.RUnlock()
+	s.bufferMutex.RLock()
+	defer s.bufferMutex.RUnlock()
+
+	return s.buffer.TotalBlocksCount()
+}
+
+func (s *Scheduler) Quanta(issuerID identity.ID) *big.Rat {
+	return big.NewRat(int64(s.getAccessMana(issuerID)), int64(s.totalAccessManaRetrieveFunc()))
+}
+
+func (s *Scheduler) Deficit(issuerID identity.ID) *big.Rat {
+	s.deficitsMutex.RLock()
+	defer s.deficitsMutex.RUnlock()
+
+	deficit, exists := s.deficits[issuerID]
+	if !exists {
+		return new(big.Rat).SetInt64(0)
+	}
+
+	return deficit
+}
+
+// Shutdown shuts down the Scheduler.
+// Shutdown blocks until the scheduler has been shutdown successfully.
+func (s *Scheduler) Shutdown() {
+	s.shutdownOnce.Do(func() {
+		s.stopped.Set()
+		s.shutdownSignal <- struct{}{}
+		close(s.shutdownSignal)
+	})
+}
+
+// Block retrieves the Block with given id from the mem-storage.
+func (s *Scheduler) Block(id models.BlockID) (block *Block, exists bool) {
+	s.EvictionManager.RLock()
+	defer s.EvictionManager.RUnlock()
+
+	return s.block(id)
+}
+
+func (s *Scheduler) AddBlock(sourceBlock *virtualvoting.Block) {
+	s.EvictionManager.RLock()
+	defer s.EvictionManager.RUnlock()
+
+	block, _ := s.getOrRegisterBlock(sourceBlock)
+
+	if err := s.Submit(block); err != nil {
+		if !errors.Is(err, ErrInsufficientMana) {
+			s.Events.Error.Trigger(errors.Wrap(err, "failed to submit to scheduler"))
+		}
+	}
+	s.tryReady(block)
+}
+
+func (s *Scheduler) HandleAcceptedBlock(acceptedBlock *acceptance.Block) {
```

Comment:
Here we still have a dependency to `acceptance`. Should better receive a `virtualvoting.Block`

Reply:
This dependency is not bad, as we can mock this block. The main motivation for not passing acceptance.Gadget to scheduler was that it would be hard to mock in tests

Reply:
Done, attached to the event in congestion control and unpacked `virtualvoting.Block`